src/ScratchStorage.js

```diff
@@ -177,49 +177,56 @@ class ScratchStorage {
-                if (helperIndex < helpers.length) {
-                    const helper = helpers[helperIndex++];
-                    helper.load(assetType, assetId, dataFormat)
-                        .then(
-                            asset => {
-                                if (asset === null) {
-                                    tryNextHelper();
-                                } else {
-                                    // TODO? this.localHelper.cache(assetType, assetId, asset);
-                                    if (helper !== this.builtinHelper && assetType.immutable) {
-                                        asset.assetId = this.builtinHelper._store(
-                                            assetType,
-                                            asset.dataFormat,
-                                            asset.data,
-                                            assetId
-                                        );
-                                    }
-                                    // Note that other attempts may have caused errors, effectively suppressed here.
-                                    resolve(asset);
-                                }
-                            },
-                            error => {
-                                errors.push(error);
-                                // TODO: maybe some types of error should prevent trying the next helper?
-                                tryNextHelper();
-                            }
-                        );
-                } else if (errors.length === 0) {
-                    // Nothing went wrong but we couldn't find the asset.
-                    resolve(null);
-                } else {
-                    // At least one thing went wrong and also we couldn't find the asset.
-                    reject(errors);
+        let helperIndex = 0;
+        let helper;
+        const tryNextHelper = err => {
+            if (err) {
+                errors.push(err);
+            }
+
+            helper = helpers[helperIndex++];
+
+            if (helper) {
+                const loading = helper.load(assetType, assetId, dataFormat);
+                if (loading === null) {
+                    return tryNextHelper();
+                }
+                return loading
+                    // TODO: maybe some types of error should prevent trying the
+                    // next helper?
+                    .catch(tryNextHelper);
+            } else if (errors.length > 0) {
+                // At least one thing went wrong and also we couldn't find the
+                // asset.
+                return Promise.reject(errors);
+            }
+
+            // Nothing went wrong but we couldn't find the asset.
+            return Promise.resolve(null);
+        };
+
+        return tryNextHelper()
+            .then(asset => {
+                if (asset === null) {
+                    return tryNextHelper();
```

Comment:
Why is this necessary when `tryNextHelper` already tries to handle null?

Reply:
You're right its not needed. Its part of [this](https://github.com/LLK/scratch-storage/pull/78/files#diff-ba10d9fbbe07eb4ce18d86eda3f7ca2eL191) original handler mostly unmodified. I'll remove it. (It might also cause an infinite loop?)

Reply:
To clarify it'll loop when it can't find anything to load.